File: services/inference/app/models/job_schemas.py
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Optional
from uuid import uuid4

from pydantic import Field

from .schemas import ApiModel


class DrawingJobStatus(str, Enum):
    QUEUED = "queued"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


class DrawingResultRowStatus(str, Enum):
    FOUND = "found"
    NOT_FOUND = "not_found"
    UNCERTAIN = "uncertain"


class ResultPoint(ApiModel):
    x: float = Field(ge=0)
    y: float = Field(ge=0)


class ResultBoundingBox(ApiModel):
    x: float = Field(ge=0)
    y: float = Field(ge=0)
    w: float = Field(gt=0)
    h: float = Field(gt=0)


class DrawingJobInput(ApiModel):
    drawing_name: str
    drawing_url: str
    labels_name: Optional[str] = None
    labels_url: Optional[str] = None
    has_labels: bool = False


class DrawingResultPage(ApiModel):
    page_index: int = Field(ge=0)
    width: int = Field(ge=1)
    height: int = Field(ge=1)
    overlay_url: Optional[str] = None
    row_count: int = Field(default=0, ge=0)
    held_back_count: int = Field(default=0, ge=0)


class DrawingResultRow(ApiModel):
    row: int = Field(ge=1)
    label: str
    page_index: int = Field(default=0, ge=0)
    center: Optional[ResultPoint] = None
    top_left: Optional[ResultPoint] = None
    bbox: Optional[ResultBoundingBox] = None
    final_score: Optional[float] = Field(default=None, ge=0, le=1)
    status: DrawingResultRowStatus
    note: Optional[str] = None
    source_kind: Optional[str] = None


class DrawingJobArtifacts(ApiModel):
    overlay_url: Optional[str] = None
    csv_url: Optional[str] = None
    xlsx_url: Optional[str] = None
    zip_url: Optional[str] = None
    review_csv_url: Optional[str] = None
    review_xlsx_url: Optional[str] = None
    review_zip_url: Optional[str] = None
    near_tie_csv_url: Optional[str] = None
    near_tie_json_url: Optional[str] = None
    source_json_url: Optional[str] = None
    result_json_url: Optional[str] = None


class DrawingJobSummary(ApiModel):
    total_rows: int = Field(default=0, ge=0)
    found_count: int = Field(default=0, ge=0)
    missing_count: int = Field(default=0, ge=0)
    uncertain_count: int = Field(default=0, ge=0)
    held_back_count: int = Field(default=0, ge=0)
    near_tie_ambiguity_count: int = Field(default=0, ge=0)
    discarded_count: int = Field(default=0, ge=0)
    document_confidence: Optional[float] = Field(default=None, ge=0, le=1)
    degraded_recognition: bool = False
    degraded_reason: Optional[str] = None
    selected_ocr_engine: Optional[str] = None
    fallback_used: bool = False
    fallback_attempted: bool = False
    fallback_failure_count: int = Field(default=0, ge=0)
    emergency_fallback_used: bool = False
    emergency_fallback_reason: Optional[str] = None
    review_recommended: bool = False
    review_reasons: list[str] = Field(default_factory=list)
    status_text: str = ""
    failure_message: Optional[str] = None


class DrawingJobResult(ApiModel):
    source_file: str
    source_labels_file: Optional[str] = None
    pages: list[DrawingResultPage] = Field(default_factory=list)
    rows: list[DrawingResultRow] = Field(default_factory=list)
    held_back_rows: list[DrawingResultRow] = Field(default_factory=list)
    missing_labels: list[str] = Field(default_factory=list)
    extra_detected_labels: list[str] = Field(default_factory=list)
    summary: DrawingJobSummary = Field(default_factory=DrawingJobSummary)
    artifacts: DrawingJobArtifacts = Field(default_factory=DrawingJobArtifacts)


class DrawingJob(ApiModel):
    job_id: str = Field(default_factory=lambda: str(uuid4()))
    title: str = "Untitled job"
    status: DrawingJobStatus = DrawingJobStatus.QUEUED
    input: DrawingJobInput
    error_message: Optional[str] = None
    result: Optional[DrawingJobResult] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class JobListItem(ApiModel):
    job_id: str
    title: str
    status: DrawingJobStatus
    drawing_name: str
    labels_name: Optional[str] = None
    document_confidence: Optional[float] = Field(default=None, ge=0, le=1)
    degraded_recognition: bool = False
    degraded_reason: Optional[str] = None
    emergency_fallback_used: bool = False
    error_message: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class CreateJobResponse(ApiModel):
    job: DrawingJob


class JobListResponse(ApiModel):
    jobs: list[JobListItem]


class BatchJobWarning(ApiModel):
    code: str
    message: str
    file_name: Optional[str] = None
    base_name: Optional[str] = None


class CreateBatchJobsResponse(ApiModel):
    batch_id: Optional[str] = None
    jobs: list[DrawingJob] = Field(default_factory=list)
    warnings: list[BatchJobWarning] = Field(default_factory=list)


class PreviewSessionResponse(ApiModel):
    session_id: str


class BatchSummary(ApiModel):
    total_jobs: int = Field(default=0, ge=0)
    queued_jobs: int = Field(default=0, ge=0)
    running_jobs: int = Field(default=0, ge=0)
    completed_jobs: int = Field(default=0, ge=0)
    failed_jobs: int = Field(default=0, ge=0)
    degraded_jobs: int = Field(default=0, ge=0)
    rescued_jobs: int = Field(default=0, ge=0)
    finished: bool = False


class DrawingJobBatch(ApiModel):
    batch_id: str = Field(default_factory=lambda: str(uuid4()))
    title: str = "Untitled batch"
    archive_name: str
    title_prefix: Optional[str] = None
    job_ids: list[str] = Field(default_factory=list)
    warnings: list[BatchJobWarning] = Field(default_factory=list)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    summary: BatchSummary = Field(default_factory=BatchSummary)


class BatchListItem(ApiModel):
    batch_id: str
    title: str
    archive_name: str
    job_count: int = Field(default=0, ge=0)
    warning_count: int = Field(default=0, ge=0)
    created_at: datetime
    updated_at: datetime
    summary: BatchSummary = Field(default_factory=BatchSummary)


class BatchListResponse(ApiModel):
    batches: list[BatchListItem] = Field(default_factory=list)


class BatchResponse(ApiModel):
    batch: DrawingJobBatch
    jobs: list[JobListItem] = Field(default_factory=list)

File: packages/shared-types/src/job-schemas.ts
import type { JsonSchema } from "./schemas";

export const drawingJobStatusSchema: JsonSchema = {
  $id: "DrawingJobStatus",
  type: "string",
  enum: ["queued", "running", "completed", "failed"],
};

export const drawingResultRowStatusSchema: JsonSchema = {
  $id: "DrawingResultRowStatus",
  type: "string",
  enum: ["found", "not_found", "uncertain"],
};

export const resultPointSchema: JsonSchema = {
  $id: "ResultPoint",
  type: "object",
  required: ["x", "y"],
  properties: {
    x: { type: "number", minimum: 0 },
    y: { type: "number", minimum: 0 },
  },
  additionalProperties: false,
};

export const resultBoundingBoxSchema: JsonSchema = {
  $id: "ResultBoundingBox",
  type: "object",
  required: ["x", "y", "w", "h"],
  properties: {
    x: { type: "number", minimum: 0 },
    y: { type: "number", minimum: 0 },
    w: { type: "number", exclusiveMinimum: 0 },
    h: { type: "number", exclusiveMinimum: 0 },
  },
  additionalProperties: false,
};

export const drawingJobInputSchema: JsonSchema = {
  $id: "DrawingJobInput",
  type: "object",
  required: ["drawingName", "drawingUrl", "labelsName", "labelsUrl", "hasLabels"],
  properties: {
    drawingName: { type: "string" },
    drawingUrl: { type: "string" },
    labelsName: { type: ["string", "null"] },
    labelsUrl: { type: ["string", "null"] },
    hasLabels: { type: "boolean" },
  },
  additionalProperties: false,
};

export const drawingResultPageSchema: JsonSchema = {
  $id: "DrawingResultPage",
  type: "object",
  required: ["pageIndex", "width", "height", "overlayUrl", "rowCount", "heldBackCount"],
  properties: {
    pageIndex: { type: "integer", minimum: 0 },
    width: { type: "integer", minimum: 1 },
    height: { type: "integer", minimum: 1 },
    overlayUrl: { type: ["string", "null"] },
    rowCount: { type: "integer", minimum: 0 },
    heldBackCount: { type: "integer", minimum: 0 },
  },
  additionalProperties: false,
};

export const drawingResultRowSchema: JsonSchema = {
  $id: "DrawingResultRow",
  type: "object",
  required: ["row", "label", "pageIndex", "center", "topLeft", "bbox", "finalScore", "status", "note", "sourceKind"],
  properties: {
    row: { type: "integer", minimum: 1 },
    label: { type: "string" },
    pageIndex: { type: "integer", minimum: 0 },
    center: { anyOf: [{ $ref: "ResultPoint" }, { type: "null" }] },
    topLeft: { anyOf: [{ $ref: "ResultPoint" }, { type: "null" }] },
    bbox: { anyOf: [{ $ref: "ResultBoundingBox" }, { type: "null" }] },
    finalScore: { type: ["number", "null"], minimum: 0, maximum: 1 },
    status: { $ref: "DrawingResultRowStatus" },
    note: { type: ["string", "null"] },
    sourceKind: { type: ["string", "null"] },
  },
  additionalProperties: false,
};

export const drawingJobArtifactsSchema: JsonSchema = {
  $id: "DrawingJobArtifacts",
  type: "object",
  required: [
    "overlayUrl",
    "csvUrl",
    "xlsxUrl",
    "zipUrl",
    "reviewCsvUrl",
    "reviewXlsxUrl",
    "reviewZipUrl",
    "nearTieCsvUrl",
    "nearTieJsonUrl",
    "sourceJsonUrl",
    "resultJsonUrl",
  ],
  properties: {
    overlayUrl: { type: ["string", "null"] },
    csvUrl: { type: ["string", "null"] },
    xlsxUrl: { type: ["string", "null"] },
    zipUrl: { type: ["string", "null"] },
    reviewCsvUrl: { type: ["string", "null"] },
    reviewXlsxUrl: { type: ["string", "null"] },
    reviewZipUrl: { type: ["string", "null"] },
    nearTieCsvUrl: { type: ["string", "null"] },
    nearTieJsonUrl: { type: ["string", "null"] },
    sourceJsonUrl: { type: ["string", "null"] },
    resultJsonUrl: { type: ["string", "null"] },
  },
  additionalProperties: false,
};

export const drawingJobSummarySchema: JsonSchema = {
  $id: "DrawingJobSummary",
  type: "object",
  required: [
    "totalRows",
    "foundCount",
    "missingCount",
    "uncertainCount",
    "heldBackCount",
    "nearTieAmbiguityCount",
    "discardedCount",
    "documentConfidence",
    "degradedRecognition",
    "degradedReason",
    "selectedOcrEngine",
    "fallbackUsed",
    "fallbackAttempted",
    "fallbackFailureCount",
    "emergencyFallbackUsed",
    "emergencyFallbackReason",
    "reviewRecommended",
    "reviewReasons",
    "statusText",
    "failureMessage",
  ],
  properties: {
    totalRows: { type: "integer", minimum: 0 },
    foundCount: { type: "integer", minimum: 0 },
    missingCount: { type: "integer", minimum: 0 },
    uncertainCount: { type: "integer", minimum: 0 },
    heldBackCount: { type: "integer", minimum: 0 },
    nearTieAmbiguityCount: { type: "integer", minimum: 0 },
    discardedCount: { type: "integer", minimum: 0 },
    documentConfidence: { type: ["number", "null"], minimum: 0, maximum: 1 },
    degradedRecognition: { type: "boolean" },
    degradedReason: { type: ["string", "null"] },
    selectedOcrEngine: { type: ["string", "null"] },
    fallbackUsed: { type: "boolean" },
    fallbackAttempted: { type: "boolean" },
    fallbackFailureCount: { type: "integer", minimum: 0 },
    emergencyFallbackUsed: { type: "boolean" },
    emergencyFallbackReason: { type: ["string", "null"] },
    reviewRecommended: { type: "boolean" },
    reviewReasons: { type: "array", items: { type: "string" } },
    statusText: { type: "string" },
    failureMessage: { type: ["string", "null"] },
  },
  additionalProperties: false,
};

export const drawingJobResultSchema: JsonSchema = {
  $id: "DrawingJobResult",
  type: "object",
  required: [
    "sourceFile",
    "sourceLabelsFile",
    "pages",
    "rows",
    "heldBackRows",
    "missingLabels",
    "extraDetectedLabels",
    "summary",
    "artifacts",
  ],
  properties: {
    sourceFile: { type: "string" },
    sourceLabelsFile: { type: ["string", "null"] },
    pages: { type: "array", items: { $ref: "DrawingResultPage" } },
    rows: { type: "array", items: { $ref: "DrawingResultRow" } },
    heldBackRows: { type: "array", items: { $ref: "DrawingResultRow" } },
    missingLabels: { type: "array", items: { type: "string" } },
    extraDetectedLabels: { type: "array", items: { type: "string" } },
    summary: { $ref: "DrawingJobSummary" },
    artifacts: { $ref: "DrawingJobArtifacts" },
  },
  additionalProperties: false,
};

export const drawingJobSchema: JsonSchema = {
  $id: "DrawingJob",
  type: "object",
  required: ["jobId", "title", "status", "input", "errorMessage", "result", "createdAt", "updatedAt"],
  properties: {
    jobId: { type: "string" },
    title: { type: "string" },
    status: { $ref: "DrawingJobStatus" },
    input: { $ref: "DrawingJobInput" },
    errorMessage: { type: ["string", "null"] },
    result: { anyOf: [{ $ref: "DrawingJobResult" }, { type: "null" }] },
    createdAt: { type: "string", format: "date-time" },
    updatedAt: { type: "string", format: "date-time" },
  },
  additionalProperties: false,
};

export const jobListItemSchema: JsonSchema = {
  $id: "JobListItem",
  type: "object",
  required: [
    "jobId",
    "title",
    "status",
    "drawingName",
    "labelsName",
    "documentConfidence",
    "degradedRecognition",
    "degradedReason",
    "emergencyFallbackUsed",
    "errorMessage",
    "createdAt",
    "updatedAt",
  ],
  properties: {
    jobId: { type: "string" },
    title: { type: "string" },
    status: { $ref: "DrawingJobStatus" },
    drawingName: { type: "string" },
    labelsName: { type: ["string", "null"] },
    documentConfidence: { type: ["number", "null"], minimum: 0, maximum: 1 },
    degradedRecognition: { type: "boolean" },
    degradedReason: { type: ["string", "null"] },
    emergencyFallbackUsed: { type: "boolean" },
    errorMessage: { type: ["string", "null"] },
    createdAt: { type: "string", format: "date-time" },
    updatedAt: { type: "string", format: "date-time" },
  },
  additionalProperties: false,
};

export const jobContracts = {
  CreateJobResponse: {
    $id: "CreateJobResponse",
    type: "object",
    required: ["job"],
    properties: {
      job: { $ref: "DrawingJob" },
    },
    additionalProperties: false,
  },
  JobListResponse: {
    $id: "JobListResponse",
    type: "object",
    required: ["jobs"],
    properties: {
      jobs: {
        type: "array",
        items: { $ref: "JobListItem" },
      },
    },
    additionalProperties: false,
  },
  BatchJobWarning: {
    $id: "BatchJobWarning",
    type: "object",
    required: ["code", "message", "fileName", "baseName"],
    properties: {
      code: { type: "string" },
      message: { type: "string" },
      fileName: { type: ["string", "null"] },
      baseName: { type: ["string", "null"] },
    },
    additionalProperties: false,
  },
  CreateBatchJobsResponse: {
    $id: "CreateBatchJobsResponse",
    type: "object",
    required: ["batchId", "jobs", "warnings"],
    properties: {
      batchId: { type: ["string", "null"] },
      jobs: {
        type: "array",
        items: { $ref: "DrawingJob" },
      },
      warnings: {
        type: "array",
        items: { $ref: "BatchJobWarning" },
      },
    },
    additionalProperties: false,
  },
  BatchSummary: {
    $id: "BatchSummary",
    type: "object",
    required: ["totalJobs", "queuedJobs", "runningJobs", "completedJobs", "failedJobs", "degradedJobs", "rescuedJobs", "finished"],
    properties: {
      totalJobs: { type: "integer", minimum: 0 },
      queuedJobs: { type: "integer", minimum: 0 },
      runningJobs: { type: "integer", minimum: 0 },
      completedJobs: { type: "integer", minimum: 0 },
      failedJobs: { type: "integer", minimum: 0 },
      degradedJobs: { type: "integer", minimum: 0 },
      rescuedJobs: { type: "integer", minimum: 0 },
      finished: { type: "boolean" },
    },
    additionalProperties: false,
  },
  DrawingJobBatch: {
    $id: "DrawingJobBatch",
    type: "object",
    required: ["batchId", "title", "archiveName", "titlePrefix", "jobIds", "warnings", "createdAt", "updatedAt", "summary"],
    properties: {
      batchId: { type: "string" },
      title: { type: "string" },
      archiveName: { type: "string" },
      titlePrefix: { type: ["string", "null"] },
      jobIds: { type: "array", items: { type: "string" } },
      warnings: { type: "array", items: { $ref: "BatchJobWarning" } },
      createdAt: { type: "string", format: "date-time" },
      updatedAt: { type: "string", format: "date-time" },
      summary: { $ref: "BatchSummary" },
    },
    additionalProperties: false,
  },
  BatchListItem: {
    $id: "BatchListItem",
    type: "object",
    required: ["batchId", "title", "archiveName", "jobCount", "warningCount", "createdAt", "updatedAt", "summary"],
    properties: {
      batchId: { type: "string" },
      title: { type: "string" },
      archiveName: { type: "string" },
      jobCount: { type: "integer", minimum: 0 },
      warningCount: { type: "integer", minimum: 0 },
      createdAt: { type: "string", format: "date-time" },
      updatedAt: { type: "string", format: "date-time" },
      summary: { $ref: "BatchSummary" },
    },
    additionalProperties: false,
  },
  BatchListResponse: {
    $id: "BatchListResponse",
    type: "object",
    required: ["batches"],
    properties: {
      batches: {
        type: "array",
        items: { $ref: "BatchListItem" },
      },
    },
    additionalProperties: false,
  },
  BatchResponse: {
    $id: "BatchResponse",
    type: "object",
    required: ["batch", "jobs"],
    properties: {
      batch: { $ref: "DrawingJobBatch" },
      jobs: {
        type: "array",
        items: { $ref: "JobListItem" },
      },
    },
    additionalProperties: false,
  },
  PreviewSessionResponse: {
    $id: "PreviewSessionResponse",
    type: "object",
    required: ["sessionId"],
    properties: {
      sessionId: { type: "string" },
    },
    additionalProperties: false,
  },
};
